### ISSUE TITLE
Supporting Clang 16 (the latest release)

### ISSUE THREAD

**Issue by andrykonchin** [davidfstr/rdiscount#155]
Clang 16 was release recently with some breaking changes. So right now building of the rdiscount native extension fails:

```
gethopt.c:48:13: error: parameter 'val' was not declared, defaults to 'int'; ISO
C99 and later do not support implicit int [-Wimplicit-int]
hopterr(ctx,val)
            ^
```

https://github.com/rubygems/rubygems/actions/runs/4847071688/jobs/8636987096

The Clang changes were handled in the Discount project [here](https://github.com/Orc/discount/commit/7afa5c62427d63177201a0f7d9254dfc6e3fca3f) (but the fix isn't release yet).

It seems updating to the newest Discount will fix the issue.

**Comment by fliiiix** [comment]
I will have a look but thats on the 3x branch and rdiscount is on the 2.x branch so either these fixes are backported to the 2.x branch or rdiscount would need to switch to 3.x (also im not sure if 3.0.0 is a real release)

(if im not misremembering switching from 2.x to 3.x is a breaking change which would require a breaking change here and some work i don't have the time to do)

**Comment by fliiiix** [comment]
Yeah see this branch here https://github.com/davidfstr/rdiscount/pull/156 this upgrade will require some work, so if you want to work on that feel free to take my branch as base there it builds and discount is updated

**Comment by eregon** [comment]
Would a minimal fix to `gethopt.c` be accepted?
The compilation error there is just because `hopterr` doesn't declare the type of `val` at all:
https://github.com/davidfstr/rdiscount/blob/aeb0b68469ce2dd236f8cc33c4c104b063bb6e8c/ext/gethopt.c#L48-L49
Fixing that is just a one-line change.
@andrykonchin Could you make a PR with that and adding TruffleRuby in CI to check it's enough?

**Comment by davidfstr** [comment]
Hi folks, I've been tracking this issue for a while. I expect to have some time to draft a fix myself on the weekend of 7/24 (in about 2 weeks).

In the meantime if someone wants to get a fix faster, please send me a PR earlier. I should have enough bandwidth to review one even before 2 weeks from now.

**Comment by eregon** [comment]
https://github.com/davidfstr/rdiscount/pull/157 fixes this

**Comment by davidfstr** [comment]
Reviewed and merged. I'll look into issuing a new RDiscount minor release later this week with the fix, v2.2.7.1

**Comment by eregon** [comment]
Thank you!

**Comment by davidfstr** [comment]
RDiscount 2.2.7.1 has been released with this fix
